modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+        // by the time we reach here, the state will never be null
+        assert state != null;
+
+        // there's a race condition between here and requestReschedule. originally this scheduleNextRun call was at the end of this
+        // block, but remember that updateDatabases can take seconds to run (it's downloading bytes from the internet), and so during the
+        // very first run there would be no future run scheduled to reschedule in requestReschedule. which meant that if you went from zero
+        // to N(>=2) databases in quick succession, then the all but the first database wouldn't necessarily get downloaded, because the
+        // requestReschedule call in the EnterpriseGeoIpDownloaderTaskExecutor's clusterChanged wouldn't have a scheduled future run to
+        // reschedule. scheduling the next run at the beginning of this run means that there's a much smaller window (milliseconds?, rather
+        // than seconds) in which such a race could occur. technically there's a window here, still, but i think it's _greatly_ reduced.
+        scheduleNextRun(pollIntervalSupplier.get());
+        // TODO regardless of the above comment, i like the idea of checking the lowest last-checked time and then running the math to get
+        // to the next interval from then -- maybe that's a neat future enhancement to add
+
+        if (isCancelled() || isCompleted()) {
+            return;
+        }
+        try {
+            updateDatabases(); // n.b. this downloads bytes from the internet, it can take a while
+        } catch (Exception e) {
+            logger.error("exception during geoip databases update", e);
+        }
+        try {
+            cleanDatabases();
+        } catch (Exception e) {
+            logger.error("exception during geoip databases cleanup", e);
+        }
+    }
+
+    /**
+     * This method requests that the downloader be rescheduled to run immediately (presumably because a dynamic property supplied by
+     * pollIntervalSupplier or eagerDownloadSupplier has changed, or a pipeline with a geoip processor has been added). This method does
+     * nothing if this task is cancelled, completed, or has not yet been scheduled to run for the first time. It cancels any existing
+     * scheduled run.
+     */
+    public void requestReschedule() {
+        if (isCancelled() || isCompleted()) {
+            return;
+        }
+        if (scheduled != null && scheduled.cancel()) {
+            scheduleNextRun(TimeValue.ZERO);
+        }
+    }
+
+    private void cleanDatabases() {
+        // this cleanDatabases logic is wrong, it'll deleteOldChunks() repeatedly forever,
+        // and it counts down the lastCheck a millisecond at a time which seems like nonsense to me.
+        List<Map.Entry<String, Metadata>> expiredDatabases = state.getDatabases()
+            .entrySet()
+            .stream()
+            .filter(e -> e.getValue().isValid(clusterService.state().metadata().settings()) == false)
```

Comment:
Can we rename the `isValid` method to something like `isNewEnough` so that it's clear from reading its use what it's checking? I had to go look it up while reading this to see what it was doing.